packages/api/lib/writeRecords/write-granules.js

```diff
@@ -299,15 +299,61 @@ const _writeGranule = async ({
   log.info('About to write granule record %j to PostgreSQL', postgresGranuleRecord);
   log.info('About to write granule record %j to DynamoDB', dynamoGranuleRecord);
 
-  await createRejectableTransaction(knex, async (trx) => {
-    granuleCumulusId = await _writePostgresGranuleViaTransaction({
-      granuleRecord: postgresGranuleRecord,
-      executionCumulusId,
-      trx,
-      granulePgModel,
+  try {
+    await createRejectableTransaction(knex, async (trx) => {
+      granuleCumulusId = await _writePostgresGranuleViaTransaction({
+        granuleRecord: postgresGranuleRecord,
+        executionCumulusId,
+        trx,
+        granulePgModel,
+      });
+      return granuleModel.storeGranule(dynamoGranuleRecord);
     });
-    return granuleModel.storeGranule(dynamoGranuleRecord);
-  });
+  } catch (thrownError) {
+    log.error(`Write Granule failed: ${JSON.stringify(thrownError)}`);
+
+    // Set status to failed if granule exists to prevent previous status
+    // from giving the wrong impression
+    const dynamoGranuleExists
+      = await granuleModel.exists({ granuleId: dynamoGranuleRecord.granuleId });
+    const postgresGranuleExists
+      = await granulePgModel.exists(
+        knex,
+        {
+          granule_id: postgresGranuleRecord.granule_id,
+          collection_cumulus_id: postgresGranuleRecord.collection_cumulus_id,
+        }
+      );
+
+    if (dynamoGranuleExists && postgresGranuleExists) {
```

Comment:
So my understanding of this code is that if an any error is caught from attempting to write a granule, then we update the granule status to "failed". However, that also could be misleading. An error could be thrown (due to database connectivity or anything) when trying to write a granule as "running" that is currently "queued", in which case immediately updating the granule to "failed" would be misleading because the granule workflow is not really in a terminal state yet.

Also, in general, this catch block could be (and most often probably will be) reached because of database connectivity issues. In that case, I'm not sure if it makes sense to immediately to try and do more database queries to update the granule.

It might be worth having an architecture discussion here about this

Reply:
Fair enough. I'll add a topic to discuss this afternoon.